pkg/storegateway/storegatewaypb/custom.go

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package storegatewaypb
+
+import (
+	"context"
+
+	"github.com/grafana/dskit/grpcutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// customStoreGatewayClient is a custom client which wraps the real gRPC client and re-maps well known
+// gRPC errors into standard golang errors.
+type customStoreGatewayClient struct {
```

Comment:
This is out of scope for this PR, but if we're going to adopt this approach everywhere, it'd be good to be able to generate these wrapped clients automatically - maintaining them manually is going to become a pain.

Reply:
Agree. Let me investigate options and I will share a proposal.

Reply:
Would be great if you could include "transform errors in grpc client" option in your proposal.

Reply:
Another option: enhance `(e *status.Error) Is(target error)` method, and let it return `true` if status code is Canceled and provided `target` is `context.Canceled`, and same for DeadlineExceeded.